File: Sorting/mergeSort.py
def mergeSort(start,end,arr):

    #base case
    if start==end:
        return
    
    mid = (start+end)//2

    # sorting the left half
    mergeSort(start,mid,arr)
    #sorting the right half
    mergeSort(mid+1,end,arr)

    #merge
    tmp1 = arr[start:(mid+1)]
    tmp2 = arr[(mid+1):(end+1)]

    l1,l2,ptr1,ptr2,ptr3 = len(tmp1),len(tmp2),0,0,start

    while ptr1<l1 and ptr2<l2:

        if tmp1[ptr1]<tmp2[ptr2]:
            arr[ptr3] = tmp1[ptr1]
            ptr3+=1
            ptr1+=1
        else:
            arr[ptr3] = tmp2[ptr2]
            ptr3+=1
            ptr2+=1
    
    while ptr1<l1:
        arr[ptr3] = tmp1[ptr1]
        ptr3+=1
        ptr1+=1
    
    while ptr2<l2:
        arr[ptr3] = tmp2[ptr2]
        ptr3+=1
        ptr2+=1





arr = [6,3,1,4,7,5]
n = len(arr)
mergeSort(0,n-1,arr)

print(arr)

File: Sorting/MergeSort.java
import java.util.*;

public class MergeSort {
    public static void main(String[] args) {
       
        ArrayList<Integer> arr  = new ArrayList<>(List.of(6,3,1,4,7,5));
        int n = arr.size();
        mergeSort(0,n-1,arr);

        for(int x: arr) System.out.print(x+" ");
        

    }

    public static void mergeSort(int start,int end,ArrayList<Integer> arr){

        //base case: single element is sorted
        if(start==end) 
            return;
        
        int mid = (start+end)/2;

        //sort the left half
        mergeSort(start,mid,arr);
        //sort the right half
        mergeSort(mid+1,end,arr);

        //merge
        ArrayList<Integer> tmp1 = new ArrayList<>();
        ArrayList<Integer> tmp2 = new ArrayList<>();

        for(int i=start;i<=end;i++){
            if(i<=mid) tmp1.add(arr.get(i));
            else tmp2.add(arr.get(i));
        }

        int l1 = tmp1.size() ,l2 = tmp2.size();
        int ptr1 =0,ptr2=0,ptr3=start;

        while(ptr1<l1 && ptr2<l2){

            if(tmp1.get(ptr1)<tmp2.get(ptr2)){
                arr.set(ptr3,tmp1.get(ptr1));
                ptr1++;
                ptr3++;
            }
            else{
                arr.set(ptr3,tmp2.get(ptr2));
                ptr2++;
                ptr3++;
            }
        }

        while(ptr1<l1){
            arr.set(ptr3,tmp1.get(ptr1));
            ptr1++;
            ptr3++;
        }

        while(ptr2<l2){
            arr.set(ptr3,tmp2.get(ptr2));
            ptr2++;
            ptr3++;           
        }
    }

}